### ISSUE TITLE
Unified concepts for `BuildOperationExecutor` and `BuildOperationProcessor`

### ISSUE THREAD

**Issue by jjohannes** [gradle/gradle#1676]
Historically, we have two overlapping ideas of build operation execution
- `BuildOperationExecutor` solely meant for _reporting_ operations (i.e. see what is going on in my build in the IDE or in build scans)
- `BuildOperationProcessor` define _units of work_ as build operations, schedule and execute them (possibly in parallel). Recent work on worker daemons have touched this area. For example: #1617 

Certain unifications and simplifications should be performed here. For example, operations that are scheduled for execution should automatically also be reported.

**Comment by eskatos** [comment]
First PR: https://github.com/gradle/gradle/pull/1699 - BuildOperationProcessor emits build operations

Remains to unify `BuildOperationExecutor` and `BuildOperationProcessor` APIs to define build operations display names, descriptors etc... This will be done as a separate PR.

**Comment by eskatos** [comment]
PR https://github.com/gradle/gradle/pull/1699 merged

**Comment by oehme** [comment]
A first step would be to use `BuildOperation` everywhere and getting rid of the `run(BuildOperationDescriptor, Action)` methods. That would also remove one layer of indirection, making the whole concept a little more memory efficient.

```
BuildOperation {
  BuildOperationDescriptor.Builder description()
}
```

```
RunnableBuildOperation extends BuildOperation{
  void run(BuildOperationContext context)
}
```

```
CallableBuildOperation<T> extends BuildOperation {
  T call(BuildOperationContext context)
}
```

Then we could unify the run single/run multiple APIs as follows:

```
BuildOperationExecutor {
  void run(RunnableBuildOperation)
  <T> T call(CallableBuildOperation<T>)
  <O extends BuildOperation> void runAll(BuildOperationWorker<O>, Action<BuildOperationQueue<O>>)
  <O extends RunnableBuildOperation> void runAll(Action<BuildOperationQueue<O>>)
}
```

```
BuildOperationQueue<T> {
  void schedule(T)
  void cancel()
  void waitForCompletion()
}
```

**Comment by oehme** [comment]
This has been merged
